src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc_NavigationBar.cs

```diff
@@ -96,5 +124,72 @@ private static UIElement GetNavbar(IWpfTextView textView)
             var vsDropDownBarAdapterMargin = topMarginControl.Content as UIElement;
             return vsDropDownBarAdapterMargin;
         }
+
+        private static IVsCodeWindow TryGetCodeWindow(IVsTextView textView)
+        {
+            if (textView == null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            if (!(textView is IObjectWithSite objectWithSite))
+            {
+                return null;
+            }
+
+            var riid = typeof(IOleServiceProvider).GUID;
+            objectWithSite.GetSite(ref riid, out var ppvSite);
+            if (ppvSite == IntPtr.Zero)
+            {
+                return null;
+            }
+
+            IOleServiceProvider oleServiceProvider = null;
+            try
+            {
+                oleServiceProvider = Marshal.GetObjectForIUnknown(ppvSite) as IOleServiceProvider;
+            }
+            finally
+            {
+                Marshal.Release(ppvSite);
+            }
+
+            if (oleServiceProvider == null)
+            {
+                return null;
+            }
+
+            var guidService = typeof(SVsWindowFrame).GUID;
+            riid = typeof(IVsWindowFrame).GUID;
+            if (ErrorHandler.Failed(oleServiceProvider.QueryService(ref guidService, ref riid, out var ppvObject)) || ppvObject == IntPtr.Zero)
+            {
+                return null;
+            }
+
+            IVsWindowFrame frame = null;
+            try
+            {
+                frame = Marshal.GetObjectForIUnknown(ppvObject) as IVsWindowFrame;
+            }
+            finally
+            {
+                Marshal.Release(ppvObject);
+            }
+
+            riid = typeof(IVsCodeWindow).GUID;
+            if (ErrorHandler.Failed(frame.QueryViewInterface(ref riid, out ppvObject)) || ppvObject == IntPtr.Zero)
+            {
+                return null;
+            }
+
+            try
+            {
+                return Marshal.GetObjectForIUnknown(ppvObject) as IVsCodeWindow;
+            }
+            finally
+            {
+                Marshal.Release(ppvObject);
+            }
```

Comment:
Consider wrapping this pattern in a helper. Seems to be used a bit.

Reply:
I'm not a fan of separating the allocation and release, and it's not easy to combine since each allocation uses a different method. I'll think about ways to improve this code though.